### DISCUSSION TITLE
[General] zks_getAllAccountBalances doesn't return full of token balances

### DISCUSSION THREAD

**Discussion by huy23121994** [zkSync-Community-Hub/zksync-developers#103] [General] [answered]
### Environment

Mainnet

### Provide a brief description of the functionality you're trying to implement.

Try to use `zks_getAllAccountBalances`
Returns all balances for confirmed tokens given by an account address.

### What steps have you already taken to try to resolve the issue?

I'm trying to get all balances of an account by calling `zks_getAllAccountBalances` but it doesn't return the full of list.
Like account `0x20f7cb624845f37b7bbf1db1af1ef71d8c5c8661`

- Checking on [https://explorer.zksync.io/address/0x20f7cb624845f37b7bbf1db1af1ef71d8c5c8661](https://explorer.zksync.io/address/0x20f7cb624845f37b7bbf1db1af1ef71d8c5c8661). I see 4 tokens, 2 tokens of them that have logo and value and the rest don't have token logo and value

![image](https://github.com/zkSync-Community-Hub/zkync-developers/assets/9770074/81bae810-6822-437e-b669-cdef14221086)

- Get balances by JSON RPC call `zks_getAllAccountBalances`, it returns only 2 records of ETH, MUTE (both have token logo, value on the explorer)

![image](https://github.com/zkSync-Community-Hub/zkync-developers/assets/9770074/580e6e66-f329-47a0-9f8f-a605e47467ab)

So I wonder, what is the difference between them (tokens that have logo and the rest)

**Comment by uF4No** [comment, marked as answer]
Hi @huy23121994 

> Returns all balances for confirmed tokens given by an account address.

Note that confirmed tokens are not just any token but tokens that are available in the official bridge. To retrieve the list of confirmed tokens you can use the RPC [zks_getConfirmedTokens method](https://era.zksync.io/docs/api/api.html#zks-getconfirmedtokens) or the `getConfirmedTokens` from the `Provider` class of the [zksync-web3 SDK](https://era.zksync.io/docs/api/js/providers.html#getconfirmedtokens).

We'll have API methods in the explorer to retrieve all the account token balances soon. For now you can use an external indexer like [Covalent](https://www.covalenthq.com/docs/networks/zksync-era/).

**Reply by bxpana** [reply]
The Etherscan API is also available now https://docs.zksync.network/api-endpoints/tokens

**Comment by bradfordbradv** [comment]
The tokens that have logos and values displayed are likely those that have been added to zkSync's token registry. Tokens added to the registry will have additional metadata associated with them, such as the token logo and value. On the other hand, tokens that do not have logos or values displayed may not be included in the token registry, hence the lack of additional metadata. To ensure that all tokens are displayed correctly, you may need to add them to zkSync's token registry if they are not already included.

**Comment by platactiomeb** [comment]
Environment: Mainnet

Description: Attempting to retrieve all balances of an account using zks_getAllAccountBalances, but only receiving balances for tokens with logos and values displayed in the explorer.

Resolution: The tokens with logos and values likely have metadata associated with them, while the others may not. Ensure that all tokens have proper metadata assigned to them to display correctly.

**Comment by lupanncoofonpowell** [comment]
Issue: When calling zks_getAllAccountBalances, only tokens with logos and values are returned for some accounts.
Status: Tokens without logos and values may not be recognized by the zkSync indexer, leading to incomplete balance data.
Solution: Ensure that all tokens are properly recognized and indexed by the zkSync indexer to fetch complete balance data.

**Comment by epjocnopasara** [comment]
The tokens that have a logo and a corresponding value are likely to be those supported by the zkSync network with established metadata, including token logo and value. On the other hand, tokens without a logo and value may be custom or lesser-known tokens that are not yet fully integrated into the zkSync network's metadata system. Integration of token metadata typically involves cooperation between zkSync and token issuers to provide accurate information about the token, including its logo and value. Therefore, the difference lies in the level of integration and metadata availability for each token on the zkSync network.

**Comment by venlaudecsubs** [comment]
The user is trying to use the zks_getAllAccountBalances function to get all balances of an account on the zkSync Mainnet, but it doesn't return the full list of tokens. Specifically, the user is seeing 4 tokens on the explorer, but only 2 tokens (ETH and MUTE) with token logo and value are returned by the zks_getAllAccountBalances function. The user is wondering what the difference is between the tokens that have logo and the ones that don't.
